#!/usr/bin/env python3
"""
PhotonRPC é›†æˆæµ‹è¯•
éªŒè¯æ•´ä¸ªRPCæ¡†æ¶çš„åŠŸèƒ½æ­£ç¡®æ€§ï¼ŒåŒ…æ‹¬æ­£å¸¸æµç¨‹å’Œè¾¹ç•Œ/é”™è¯¯æƒ…å†µ
"""

import socket
import struct
import subprocess
import time
import threading
import os
import sys
import signal
from typing import Optional

import rpc_message_pb2
import echo_service_pb2
import calculate_service_pb2

# æµ‹è¯•é…ç½®
SERVER_HOST = "127.0.0.1"
SERVER_PORT = 12345
TEST_TIMEOUT = 5  # ç§’

# Codecå®ç°ï¼ˆä¸C++ç‰ˆæœ¬ä¸€è‡´ï¼‰
class Codec:
    """æ¶ˆæ¯ç¼–è§£ç å™¨ï¼Œå®ç°ä¸C++ç‰ˆæœ¬ä¸€è‡´çš„åè®®"""
    
    @staticmethod
    def encode(data: bytes) -> bytes:
        """ç¼–ç ï¼š4å­—èŠ‚é•¿åº¦å‰ç¼€ + æ•°æ®"""
        length = len(data)
        length_bytes = struct.pack('<I', length)  # å°ç«¯åºï¼Œ4å­—èŠ‚
        return length_bytes + data
    
    @staticmethod
    def decode(encoded_data: bytes) -> bytes:
        """è§£ç ï¼šè¯»å–4å­—èŠ‚é•¿åº¦å‰ç¼€ï¼Œè¿”å›æ•°æ®éƒ¨åˆ†"""
        if len(encoded_data) < 4:
            return b''
        length = struct.unpack('<I', encoded_data[:4])[0]
        if len(encoded_data) < 4 + length:
            return b''
        return encoded_data[4:4+length]


# Protobufæ¶ˆæ¯æ„é€ ï¼ˆä½¿ç”¨ç”Ÿæˆçš„protobufæ¨¡å—ï¼‰

def build_rpc_message(message_id: int, message_type: int, 
                     service_name: str, method_name: str,
                     request_data: bytes = b'') -> bytes:
    """æ„é€ RpcMessage protobufæ¶ˆæ¯"""
    rpc_msg = rpc_message_pb2.RpcMessage()
    rpc_msg.id = message_id
    rpc_msg.type = message_type
    rpc_msg.service_name = service_name
    rpc_msg.method_name = method_name
    
    if message_type == rpc_message_pb2.MessageType.RPC_TYPE_REQUEST and request_data:
        rpc_msg.request = request_data
    elif message_type == rpc_message_pb2.MessageType.RPC_TYPE_RESPONSE and request_data:
        rpc_msg.response = request_data
    
    return rpc_msg.SerializeToString()


def parse_rpc_response(data: bytes) -> Optional[bytes]:
    """è§£æRPCå“åº”æ¶ˆæ¯ï¼Œè¿”å›responseå­—æ®µå†…å®¹"""
    rpc_msg = rpc_message_pb2.RpcMessage()
    rpc_msg.ParseFromString(data)
    return rpc_msg.response if rpc_msg.response else None


# Echoå’ŒCalculateæœåŠ¡çš„protobufæ¶ˆæ¯æ„é€ 
def build_echo_request(sentence: str) -> bytes:
    """æ„é€ EchoRequest: sentence (field 1, string)"""
    echo_req = echo_service_pb2.EchoRequest()
    echo_req.sentence = sentence
    return echo_req.SerializeToString()


def parse_echo_response(data: bytes) -> Optional[str]:
    """è§£æEchoResponse: result (field 1, string)"""
    echo_resp = echo_service_pb2.EchoResponse()
    echo_resp.ParseFromString(data)
    return echo_resp.result


def build_add_request(a: int, b: int) -> bytes:
    """æ„é€ AddRequest: a (field 1, int32), b (field 2, int32)"""
    add_req = calculate_service_pb2.AddRequest()
    add_req.a = a
    add_req.b = b
    return add_req.SerializeToString()


def build_sub_request(a: int, b: int) -> bytes:
    """æ„é€ SubRequest: a (field 1, int32), b (field 2, int32)"""
    sub_req = calculate_service_pb2.SubRequest()
    sub_req.a = a
    sub_req.b = b
    return sub_req.SerializeToString()


def parse_add_response(data: bytes) -> Optional[int]:
    """è§£æAddResponse: result (field 1, int32)"""
    add_resp = calculate_service_pb2.AddResponse()
    add_resp.ParseFromString(data)
    return add_resp.result


def parse_sub_response(data: bytes) -> Optional[int]:
    """è§£æSubResponse: result (field 1, int32)"""
    sub_resp = calculate_service_pb2.SubResponse()
    sub_resp.ParseFromString(data)
    return sub_resp.result


# RPCå®¢æˆ·ç«¯ç±»
class RpcClient:
    """RPCå®¢æˆ·ç«¯ï¼Œç”¨äºå‘é€è¯·æ±‚å’Œæ¥æ”¶å“åº”"""
    
    def __init__(self, host: str = SERVER_HOST, port: int = SERVER_PORT):
        self.host = host
        self.port = port
        self.sock = None
    
    def connect(self) -> bool:
        """è¿æ¥åˆ°æœåŠ¡å™¨"""
        try:
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.sock.settimeout(TEST_TIMEOUT)
            self.sock.connect((self.host, self.port))
            return True
        except Exception as e:
            print(f"è¿æ¥å¤±è´¥: {e}")
            return False
    
    def call_rpc(self, service_name: str, method_name: str, 
                 request_data: bytes, message_id: int = 1) -> Optional[bytes]:
        """å‘é€RPCè¯·æ±‚å¹¶è¿”å›å“åº”æ•°æ®"""
        if not self.sock:
            if not self.connect():
                return None
        
        try:
            # æ„é€ RPCæ¶ˆæ¯
            rpc_message = build_rpc_message(
                message_id, 
                rpc_message_pb2.MessageType.RPC_TYPE_REQUEST,
                service_name,
                method_name,
                request_data
            )
            
            # ç¼–ç å¹¶å‘é€
            encoded = Codec.encode(rpc_message)
            self.sock.sendall(encoded)
            
            # æ¥æ”¶å“åº”
            # å…ˆè¯»å–4å­—èŠ‚é•¿åº¦
            length_data = b''
            while len(length_data) < 4:
                chunk = self.sock.recv(4 - len(length_data))
                if not chunk:
                    return None
                length_data += chunk
            
            length = struct.unpack('<I', length_data)[0]
            
            # è¯»å–æ¶ˆæ¯ä½“
            message_data = b''
            while len(message_data) < length:
                chunk = self.sock.recv(length - len(message_data))
                if not chunk:
                    return None
                message_data += chunk
            
            # è§£æå“åº”
            response_data = parse_rpc_response(message_data)
            return response_data
            
        except socket.timeout:
            print(f"è¯·æ±‚è¶…æ—¶")
            return None
        except Exception as e:
            print(f"RPCè°ƒç”¨å¤±è´¥: {e}")
            return None
    
    def close(self):
        """å…³é—­è¿æ¥"""
        if self.sock:
            self.sock.close()
            self.sock = None


# æµ‹è¯•è¾…åŠ©å‡½æ•°
class TestHelper:
    """æµ‹è¯•è¾…åŠ©ç±»"""
    
    @staticmethod
    def wait_for_server(host: str, port: int, timeout: int = 10) -> bool:
        """ç­‰å¾…æœåŠ¡å™¨å¯åŠ¨"""
        end_time = time.time() + timeout
        while time.time() < end_time:
            try:
                sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                sock.settimeout(1)
                result = sock.connect_ex((host, port))
                sock.close()
                if result == 0:
                    return True
            except:
                pass
            time.sleep(0.1)
        return False
    
    @staticmethod
    def start_server(server_path: str) -> Optional[subprocess.Popen]:
        """å¯åŠ¨æœåŠ¡å™¨è¿›ç¨‹"""
        try:
            # åˆ‡æ¢åˆ°æ­£ç¡®çš„ç›®å½•ä»¥ç¡®ä¿é…ç½®æ–‡ä»¶è·¯å¾„æ­£ç¡®
            test_dir = os.path.dirname(os.path.abspath(__file__))
            project_root = os.path.dirname(os.path.dirname(test_dir))
            server_full_path = os.path.join(project_root, server_path)
            
            if not os.path.exists(server_full_path):
                print(f"æœåŠ¡å™¨ç¨‹åºä¸å­˜åœ¨: {server_full_path}")
                return None
            
            process = subprocess.Popen(
                [server_full_path],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                cwd=project_root
            )
            return process
        except Exception as e:
            print(f"å¯åŠ¨æœåŠ¡å™¨å¤±è´¥: {e}")
            return None
    
    @staticmethod
    def stop_server(process: Optional[subprocess.Popen]):
        """åœæ­¢æœåŠ¡å™¨è¿›ç¨‹"""
        if process:
            try:
                process.terminate()
                try:
                    process.wait(timeout=2)
                except subprocess.TimeoutExpired:
                    process.kill()
                    process.wait()
            except:
                pass


# ============================================================================
# æµ‹è¯•ç”¨ä¾‹
# ============================================================================

def test_normal_echo_call():
    """æµ‹è¯•1: æ­£å¸¸EchoServiceè°ƒç”¨"""
    print("\n[æµ‹è¯•1] æ­£å¸¸EchoServiceè°ƒç”¨")
    client = RpcClient()
    try:
        request_data = build_echo_request("Hello, PhotonRPC!")
        response_data = client.call_rpc("EchoService", "Echo", request_data)
        
        assert response_data is not None, "å“åº”ä¸åº”ä¸ºç©º"
        result = parse_echo_response(response_data)
        assert result == "Hello, PhotonRPC!", f"æœŸæœ› 'Hello, PhotonRPC!'ï¼Œå¾—åˆ° '{result}'"
        print("âœ“ EchoServiceè°ƒç”¨æˆåŠŸ")
    finally:
        client.close()


def test_normal_calculate_add():
    """æµ‹è¯•2: æ­£å¸¸CalculateService.Addè°ƒç”¨"""
    print("\n[æµ‹è¯•2] æ­£å¸¸CalculateService.Addè°ƒç”¨")
    client = RpcClient()
    try:
        request_data = build_add_request(5, 6)
        response_data = client.call_rpc("CalculateService", "Add", request_data)
        
        assert response_data is not None, "å“åº”ä¸åº”ä¸ºç©º"
        result = parse_add_response(response_data)
        assert result == 11, f"æœŸæœ› 11ï¼Œå¾—åˆ° {result}"
        print("âœ“ CalculateService.Addè°ƒç”¨æˆåŠŸ")
    finally:
        client.close()


def test_normal_calculate_sub():
    """æµ‹è¯•3: æ­£å¸¸CalculateService.Subè°ƒç”¨"""
    print("\n[æµ‹è¯•3] æ­£å¸¸CalculateService.Subè°ƒç”¨")
    client = RpcClient()
    try:
        request_data = build_sub_request(10, 3)
        response_data = client.call_rpc("CalculateService", "Sub", request_data)
        
        assert response_data is not None, "å“åº”ä¸åº”ä¸ºç©º"
        result = parse_sub_response(response_data)
        assert result == 7, f"æœŸæœ› 7ï¼Œå¾—åˆ° {result}"
        print("âœ“ CalculateService.Subè°ƒç”¨æˆåŠŸ")
    finally:
        client.close()


def test_empty_string():
    """æµ‹è¯•4: ç©ºå­—ç¬¦ä¸²å‚æ•°"""
    print("\n[æµ‹è¯•4] ç©ºå­—ç¬¦ä¸²å‚æ•°")
    client = RpcClient()
    try:
        request_data = build_echo_request("")
        response_data = client.call_rpc("EchoService", "Echo", request_data)
        
        assert response_data is not None, "å“åº”ä¸åº”ä¸ºç©º"
        result = parse_echo_response(response_data)
        assert result == "", f"æœŸæœ›ç©ºå­—ç¬¦ä¸²ï¼Œå¾—åˆ° '{result}'"
        print("âœ“ ç©ºå­—ç¬¦ä¸²å¤„ç†æˆåŠŸ")
    finally:
        client.close()


def test_large_string():
    """æµ‹è¯•5: å¤§å­—ç¬¦ä¸²å‚æ•°"""
    print("\n[æµ‹è¯•5] å¤§å­—ç¬¦ä¸²å‚æ•°")
    client = RpcClient()
    try:
        large_string = "A" * 10000  # 10KBå­—ç¬¦ä¸²
        request_data = build_echo_request(large_string)
        response_data = client.call_rpc("EchoService", "Echo", request_data)
        
        assert response_data is not None, "å“åº”ä¸åº”ä¸ºç©º"
        result = parse_echo_response(response_data)
        assert result == large_string, "å¤§å­—ç¬¦ä¸²åº”å®Œæ•´è¿”å›"
        print("âœ“ å¤§å­—ç¬¦ä¸²å¤„ç†æˆåŠŸ")
    finally:
        client.close()


def test_negative_numbers():
    """æµ‹è¯•6: è´Ÿæ•°è®¡ç®—"""
    print("\n[æµ‹è¯•6] è´Ÿæ•°è®¡ç®—")
    client = RpcClient()
    try:
        # æµ‹è¯•è´Ÿæ•°åŠ æ³•ï¼š(-5) + 3 = -2
        # æ³¨æ„ï¼šè¿™é‡Œä½¿ç”¨ç®€åŒ–å®ç°ï¼Œå¯èƒ½éœ€è¦è°ƒæ•´
        request_data = build_add_request(-5, 3)
        response_data = client.call_rpc("CalculateService", "Add", request_data)
        
        if response_data:
            result = parse_add_response(response_data)
            # å¦‚æœè§£ææˆåŠŸï¼ŒéªŒè¯ç»“æœ
            if result is not None:
                print(f"âœ“ è´Ÿæ•°è®¡ç®—è¿”å›: {result}")
        else:
            print("! è´Ÿæ•°è®¡ç®—å¯èƒ½ä¸è¢«æ”¯æŒï¼ˆç®€åŒ–å®ç°é™åˆ¶ï¼‰")
    finally:
        client.close()


def test_concurrent_requests():
    """æµ‹è¯•7: å¹¶å‘è¯·æ±‚"""
    print("\n[æµ‹è¯•7] å¹¶å‘è¯·æ±‚")
    
    def make_request(thread_id: int):
        client = RpcClient()
        try:
            request_data = build_echo_request(f"Thread-{thread_id}")
            response_data = client.call_rpc("EchoService", "Echo", request_data)
            if response_data:
                result = parse_echo_response(response_data)
                return result == f"Thread-{thread_id}"
            return False
        finally:
            client.close()
    
    threads = []
    results = []
    
    def worker(thread_id):
        results.append(make_request(thread_id))
    
    for i in range(10):
        t = threading.Thread(target=worker, args=(i,))
        threads.append(t)
        t.start()
    
    for t in threads:
        t.join()
    
    success_count = sum(results)
    assert success_count == 10, f"æœŸæœ›10ä¸ªæˆåŠŸï¼Œå¾—åˆ°{success_count}ä¸ª"
    print(f"âœ“ å¹¶å‘è¯·æ±‚æˆåŠŸ ({success_count}/10)")


def test_invalid_service_name():
    """æµ‹è¯•8: æ— æ•ˆçš„æœåŠ¡å"""
    print("\n[æµ‹è¯•8] æ— æ•ˆçš„æœåŠ¡å")
    client = RpcClient()
    try:
        request_data = build_echo_request("test")
        response_data = client.call_rpc("NonExistentService", "Echo", request_data)
        # æœåŠ¡å™¨å¯èƒ½ä¼šå´©æºƒæˆ–è¿”å›é”™è¯¯ï¼ˆå–å†³äºå®ç°ï¼‰
        # è¿™é‡Œæˆ‘ä»¬ä¸»è¦éªŒè¯ä¸ä¼šå¯¼è‡´å®¢æˆ·ç«¯å´©æºƒ
        print("! æ— æ•ˆæœåŠ¡åæµ‹è¯•å®Œæˆï¼ˆæœåŠ¡å™¨å¯èƒ½å´©æºƒæˆ–è¿”å›é”™è¯¯ï¼‰")
    except Exception as e:
        print(f"! æ•è·åˆ°å¼‚å¸¸ï¼ˆé¢„æœŸè¡Œä¸ºï¼‰: {e}")
    finally:
        client.close()


def test_invalid_method_name():
    """æµ‹è¯•9: æ— æ•ˆçš„æ–¹æ³•å"""
    print("\n[æµ‹è¯•9] æ— æ•ˆçš„æ–¹æ³•å")
    client = RpcClient()
    try:
        request_data = build_echo_request("test")
        response_data = client.call_rpc("EchoService", "NonExistentMethod", request_data)
        # æœåŠ¡å™¨å¯èƒ½ä¼šå´©æºƒæˆ–è¿”å›é”™è¯¯
        print("! æ— æ•ˆæ–¹æ³•åæµ‹è¯•å®Œæˆï¼ˆæœåŠ¡å™¨å¯èƒ½å´©æºƒæˆ–è¿”å›é”™è¯¯ï¼‰")
    except Exception as e:
        print(f"! æ•è·åˆ°å¼‚å¸¸ï¼ˆé¢„æœŸè¡Œä¸ºï¼‰: {e}")
    finally:
        client.close()


def test_invalid_protobuf_data():
    """æµ‹è¯•10: æ— æ•ˆçš„protobufæ•°æ®"""
    print("\n[æµ‹è¯•10] æ— æ•ˆçš„protobufæ•°æ®")
    client = RpcClient()
    try:
        invalid_data = b"invalid protobuf data\x00\x01\x02"
        response_data = client.call_rpc("EchoService", "Echo", invalid_data)
        # æœåŠ¡å™¨åº”è¯¥èƒ½å¤„ç†æ— æ•ˆæ•°æ®è€Œä¸å´©æºƒ
        print("! æ— æ•ˆprotobufæ•°æ®æµ‹è¯•å®Œæˆï¼ˆæœåŠ¡å™¨å¯èƒ½è¿”å›é”™è¯¯æˆ–å´©æºƒï¼‰")
    except Exception as e:
        print(f"! æ•è·åˆ°å¼‚å¸¸ï¼ˆé¢„æœŸè¡Œä¸ºï¼‰: {e}")
    finally:
        client.close()


def test_multiple_sequential_calls():
    """æµ‹è¯•11: å¤šæ¬¡é¡ºåºè°ƒç”¨ï¼ˆè¿æ¥å¤ç”¨ï¼‰"""
    print("\n[æµ‹è¯•11] å¤šæ¬¡é¡ºåºè°ƒç”¨")
    client = RpcClient()
    try:
        # ç¬¬ä¸€æ¬¡è°ƒç”¨
        request_data1 = build_echo_request("First")
        response_data1 = client.call_rpc("EchoService", "Echo", request_data1)
        assert response_data1 is not None
        result1 = parse_echo_response(response_data1)
        assert result1 == "First"
        
        # ç¬¬äºŒæ¬¡è°ƒç”¨ï¼ˆä½¿ç”¨åŒä¸€è¿æ¥ï¼‰
        request_data2 = build_add_request(7, 8)
        response_data2 = client.call_rpc("CalculateService", "Add", request_data2)
        assert response_data2 is not None
        result2 = parse_add_response(response_data2)
        assert result2 == 15
        
        # ç¬¬ä¸‰æ¬¡è°ƒç”¨
        request_data3 = build_echo_request("Third")
        response_data3 = client.call_rpc("EchoService", "Echo", request_data3)
        assert response_data3 is not None
        result3 = parse_echo_response(response_data3)
        assert result3 == "Third"
        
        print("âœ“ å¤šæ¬¡é¡ºåºè°ƒç”¨æˆåŠŸ")
    finally:
        client.close()


def test_connection_refused():
    """æµ‹è¯•12: æœåŠ¡å™¨æœªå¯åŠ¨æ—¶çš„è¿æ¥é”™è¯¯"""
    print("\n[æµ‹è¯•12] è¿æ¥è¢«æ‹’ç»ï¼ˆæœåŠ¡å™¨æœªå¯åŠ¨ï¼‰")
    # ä½¿ç”¨ä¸€ä¸ªä¸å­˜åœ¨çš„ç«¯å£
    client = RpcClient(SERVER_HOST, 99999)
    try:
        request_data = build_echo_request("test")
        response_data = client.call_rpc("EchoService", "Echo", request_data)
        # åº”è¯¥è¿”å›Noneæˆ–æŠ›å‡ºå¼‚å¸¸
        assert response_data is None or False, "è¿æ¥åº”è¯¥å¤±è´¥"
        print("âœ“ è¿æ¥é”™è¯¯å¤„ç†æ­£ç¡®")
    except Exception as e:
        print(f"âœ“ æ•è·åˆ°è¿æ¥å¼‚å¸¸ï¼ˆé¢„æœŸè¡Œä¸ºï¼‰: {type(e).__name__}")
    finally:
        client.close()


def test_special_characters():
    """æµ‹è¯•13: ç‰¹æ®Šå­—ç¬¦å¤„ç†"""
    print("\n[æµ‹è¯•13] ç‰¹æ®Šå­—ç¬¦å¤„ç†")
    client = RpcClient()
    try:
        special_strings = [
            "Hello\nWorld",
            "Tab\tHere",
            "ä¸­æ–‡æµ‹è¯•",
            "ğŸš€ Emoji Test",
            "Null\x00Byte",
            "Special: !@#$%^&*()"
        ]
        
        for special_str in special_strings:
            request_data = build_echo_request(special_str)
            response_data = client.call_rpc("EchoService", "Echo", request_data)
            if response_data:
                result = parse_echo_response(response_data)
                # æŸäº›å­—ç¬¦å¯èƒ½åœ¨ä¼ è¾“ä¸­ä¸¢å¤±ï¼Œè¿™æ˜¯å¯ä»¥æ¥å—çš„
                print(f"  è¾“å…¥: {repr(special_str[:20])}, è¾“å‡º: {repr(result[:20] if result else None)}")
        
        print("âœ“ ç‰¹æ®Šå­—ç¬¦æµ‹è¯•å®Œæˆ")
    finally:
        client.close()


# ============================================================================
# ä¸»æµ‹è¯•å‡½æ•°
# ============================================================================

def run_all_tests():
    """è¿è¡Œæ‰€æœ‰æµ‹è¯•"""
    print("=" * 60)
    print("PhotonRPC é›†æˆæµ‹è¯•")
    print("=" * 60)
    
    # æŸ¥æ‰¾æœåŠ¡å™¨ç¨‹åº
    test_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.dirname(os.path.dirname(test_dir))
    server_path = os.path.join(project_root, "bin", "TestProvider")
    
    if not os.path.exists(server_path):
        # å°è¯•å…¶ä»–å¯èƒ½çš„ä½ç½®
        server_path = os.path.join(project_root, "build", "bin", "TestProvider")
        if not os.path.exists(server_path):
            print(f"é”™è¯¯: æ‰¾ä¸åˆ°æœåŠ¡å™¨ç¨‹åº TestProvider")
            print(f"è¯·åœ¨ {project_root}/bin/ æˆ– {project_root}/build/bin/ ä¸­æ„å»ºæœåŠ¡å™¨")
            return 1
    
    # å¯åŠ¨æœåŠ¡å™¨
    print(f"\nå¯åŠ¨æœåŠ¡å™¨: {server_path}")
    server_process = TestHelper.start_server("bin/TestProvider")
    if not server_process:
        print("é”™è¯¯: æ— æ³•å¯åŠ¨æœåŠ¡å™¨")
        return 1
    
    try:
        # ç­‰å¾…æœåŠ¡å™¨å¯åŠ¨
        print("ç­‰å¾…æœåŠ¡å™¨å¯åŠ¨...")
        if not TestHelper.wait_for_server(SERVER_HOST, SERVER_PORT, timeout=10):
            print("é”™è¯¯: æœåŠ¡å™¨å¯åŠ¨è¶…æ—¶")
            return 1
        
        print("æœåŠ¡å™¨å·²å¯åŠ¨ï¼Œå¼€å§‹æµ‹è¯•\n")
        
        # è¿è¡Œæµ‹è¯•
        tests = [
            ("æ­£å¸¸æµç¨‹æµ‹è¯•", [
                test_normal_echo_call,
                test_normal_calculate_add,
                test_normal_calculate_sub,
                test_multiple_sequential_calls,
            ]),
            ("è¾¹ç•Œæƒ…å†µæµ‹è¯•", [
                test_empty_string,
                test_large_string,
                test_negative_numbers,
                test_special_characters,
                test_concurrent_requests,
            ]),
            ("é”™è¯¯å¤„ç†æµ‹è¯•", [
                test_connection_refused,
                test_invalid_service_name,
                test_invalid_method_name,
                test_invalid_protobuf_data,
            ]),
        ]
        
        passed = 0
        failed = 0
        
        for category, test_list in tests:
            print(f"\n{'='*60}")
            print(f"{category}")
            print(f"{'='*60}")
            
            for test_func in test_list:
                try:
                    test_func()
                    passed += 1
                except AssertionError as e:
                    print(f"âœ— æµ‹è¯•å¤±è´¥: {e}")
                    failed += 1
                except Exception as e:
                    print(f"âœ— æµ‹è¯•å¼‚å¸¸: {type(e).__name__}: {e}")
                    failed += 1
                time.sleep(0.1)  # çŸ­æš‚å»¶è¿Ÿé¿å…è¿‡å¿«
        
        # æ±‡æ€»
        print(f"\n{'='*60}")
        print("æµ‹è¯•æ±‡æ€»")
        print(f"{'='*60}")
        print(f"é€šè¿‡: {passed}")
        print(f"å¤±è´¥: {failed}")
        print(f"æ€»è®¡: {passed + failed}")
        
        return 0 if failed == 0 else 1
        
    finally:
        # åœæ­¢æœåŠ¡å™¨
        print("\nåœæ­¢æœåŠ¡å™¨...")
        TestHelper.stop_server(server_process)
        print("æµ‹è¯•å®Œæˆ")


if __name__ == "__main__":
    sys.exit(run_all_tests())
